archive/imo/imo1960_q1.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2020 Kevin Lacker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Lacker
+-/
+
+import data.nat.digits
+
+open nat
+
+/-!
+# IMO 1960 Q1
+
+Determine all three-digit numbers $N$ having the property that $N$ is divisible by 11, and
+$\dfrac{N}{11}$ is equal to the sum of the squares of the digits of $N$.
+
+Since Lean doesn't have a way to directly express problem statements of the form
+"Determine all X satisfying Y", we express two predicates where proving that one implies the
+other is equivalent to solving the problem. A human solver also has to discover the
+second predicate.
+
+In theory, we should be able to brute force this and check every number up to 1000.
+In practice, the Lean tactics that are capable of checking a single number, like `norm_num`,
+are not efficient enough to run a thousand times while compiling one file.
+-/
+
+def sum_of_squares (L : list ℕ) : ℕ := L.foldr (λ x y, x*x + y) 0
+
+def problem_predicate (n : ℕ) :=
+(nat.digits 10 n).length = 3 ∧ 11 ∣ n ∧ n / 11 = sum_of_squares (nat.digits 10 n)
```

Comment:
```lean
(nat.digits 10 n).length = 3 ∧ n = 11 * sum_of_squares (nat.digits 10 n)
```

is it any faster/better if reformulated like this? or in general, would shuffling the order of the three make a difference (evaluation order/short circuiting?)